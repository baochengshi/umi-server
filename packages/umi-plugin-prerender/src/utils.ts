export const isDynamicRoute = (path: string): boolean =>
  !!path?.split('/')?.some?.(snippet => snippet.startsWith(':'));

export const removeSuffixHtml = (path: string): string =>
  path
    .replace('?', '')
    .replace('(', '')
    .replace(')', '')
    .replace(/\.(html|htm)/g, '');

const isHtmlPath = (path: string): boolean => /\.(html|htm)/g.test(path);

export const findJSON = (baseDir, fileName) => {
  const { join } = require('path');
  const { existsSync } = require('fs');
  const absFilePath = join(baseDir, fileName);
  if (existsSync(absFilePath)) {
    return absFilePath;
  }
  return '';
};

export const fixHtmlSuffix = route => {
  if (route.path && route.path !== '/' && !isHtmlPath(route.path) && !route.redirect) {
    route.path = `${route.path}(.html)?`;
  }
};

export const getStaticRoutePaths = (_, routes) =>
  _.uniq(
    routes.reduce((memo, route) => {
      if (route.path && !route.redirect) {
        memo.push(removeSuffixHtml(route.path));
        if (route.routes) {
          memo = memo.concat(getStaticRoutePaths(_, route.routes));
        }
      }
      return memo;
    }, []),
  );

/**
 * convert route path into file path
 * / => index.html
 * /a/b => a/b.html
 * /a/:id => a/[id].html
 * /a/b/:id/:id => a/b/[id]/[id].html
 *
 * @param path
 */
export const routeToFile = (path: string): string => {
  const pathArr = path?.split('/')?.map?.(p => {
    const normalPath = removeSuffixHtml(p);
    return isDynamicRoute(normalPath) ? `[${normalPath.replace(/:/g, '')}]` : normalPath;
  });
  const pathname = pathArr.slice(1).join('/');
  return pathname ? `${pathname}.html` : 'index.html';
};
