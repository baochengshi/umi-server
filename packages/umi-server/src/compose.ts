import { Handler } from './index';

/**
 * render html handlers pipe
 * const resultHtml = compose(
 *   handler1,
 *   handler2,
 *   hanlder3,
 *   ...
 * )($, args);
 *
 * @param handler ($, args) => $
 */
const compose: (...handler: Handler[]) => Handler = (...handler) =>
  handler.reduce((acc, curr) => ($, args) => curr(acc($, args) || $, args) || $);

export default compose;
