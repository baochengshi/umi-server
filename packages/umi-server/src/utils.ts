/* eslint-disable @typescript-eslint/no-unused-vars */
import isPlainObject from 'lodash/isPlainObject';
import merge from 'lodash/merge';
import { parse } from 'url';
import { load } from 'cheerio';
import { Handler, RenderOpts, FilterContext } from './index';

type FilterRootContainer = (
  ssrHtml: string,
  functor?: (html: string, context?: FilterContext) => string,
) => string;
/**
 * root html fragment string not parse by cheerio for better perfs
 * 1. <body>(.*)</body> => <body><!-- UMI_SERVER_TMP_PLACEHOLDER --></body>
 * 2. handlers => postProcessHtml html fragment
 * 3. <body><!-- UMI_SERVER_TMP_PLACEHOLDER -->...handler created</body>
 *    => <body>(.*)...handler created</body>
 *
 * @param html React server render origin html string
 * @param functor layout html functor for hanlders using cheerio
 */
export const filterRootContainer: FilterRootContainer = (html, functor) => {
  const bodyExp = /<body>(.*?)<\/body>/is;
  const placeholderExp = /<!-- UMI_SERVER_TMP_PLACEHOLDER -->/gs;
  const placeholder = '<body><!-- UMI_SERVER_TMP_PLACEHOLDER --></body>';
  const layout = html.replace(bodyExp, placeholder);
  const root = html.match(bodyExp)?.[1] || '';
  const matchPublicPath = root.match(/<script.*src="([^"]*)\/?umi\.(\w+\.)?js"[^>]*>/i);
  const publicPath = matchPublicPath?.[1] || '/';
  const context = {
    publicPath,
  };
  const layoutHtml = functor(layout, context);
  return layoutHtml.replace(placeholderExp, root);
};

export const _getDocumentHandler: typeof load = (html, option = {}) => {
  const docTypeHtml = /^<!DOCTYPE html>/.test(html) ? html : `<!DOCTYPE html>${html}`;
  return load(docTypeHtml, {
    decodeEntities: false,
    recognizeSelfClosing: true,
    ...option,
  });
};

export const injectChunkMaps: Handler = ($, args) => {
  const { chunkMap, publicPath = '/' } = args;
  const { js = [], css = [] } = chunkMap || {};
  // filter umi.css and umi.*.css, htmlMap have includes
  const styles = css.filter(style => !/^umi\.(\w+\.)?css$/g.test(style)) || [];
  styles.forEach(style => {
    if (style) {
      $('head').append(`<link rel="stylesheet" href="${publicPath}${style}" />`);
    }
  });
  // filter umi.js and umi.*.js
  const scripts = js.filter(script => !/^umi([.\w]*)?\.js$/g.test(script)) || [];
  scripts.forEach(script => {
    if (script) {
      $('head').append(`<link rel="preload" href="${publicPath}${script}" as="script"/>`);
    }
  });

  return $;
};

export type INodePolyfillDecorator = (
  enable: boolean,
  url?: string,
) => (renderOpts?: RenderOpts, context?: { url: string }) => void;

export const nodePolyfillDecorator: INodePolyfillDecorator = (
  enable = false,
  origin = 'http://localhost',
) => {
  const { pathname: defaultPathname } = parse(origin);
  // @ts-ignore
  global.window = {};
  if (enable) {
    const ssrPolyfill = require('ssr-polyfill');
    const mockWin = ssrPolyfill({
      url: origin,
    });
    const mountGlobal = ['document', 'location', 'navigator', 'Image', 'self'];
    mountGlobal.forEach(mount => {
      global[mount] = mockWin[mount];
    });
    // @ts-ignore
    global.window = mockWin;

    // using window.document, window.location to mock document, location
    mountGlobal.forEach(mount => {
      global[mount] = mockWin[mount];
    });

    // if use pathname to mock location.pathname
    return (
      renderOpts,
      context = {
        url: defaultPathname,
      },
    ) => {
      const { url } = context;
      const { polyfill, runInMockContext } = renderOpts;
      let nextOrigin = url;
      if ((polyfill as any)?.host) {
        nextOrigin = `${(polyfill as any).host}${url}`;
      }
      const { protocol, host } = parse(origin);
      const nextUrl = /^https?:\/\//.test(nextOrigin) ? nextOrigin : `${protocol}//${host}${url}`;
      const nextObj = parse(nextUrl);
      // @ts-ignore
      Object.defineProperty(global.window, 'location', {
        writable: true,
        value: {
          // patch window.location.origin
          origin: `${nextObj.protocol}//${nextObj.hostname}${nextObj?.port || ''}`,
          ...nextObj,
        },
      });
      // @ts-ignore
      global.location = {
        // patch location.origin
        origin: `${nextObj.protocol}//${nextObj.hostname}${nextObj?.port || ''}`,
        ...nextObj,
      };
      if (runInMockContext) {
        let mockContext;
        if (isPlainObject(runInMockContext)) {
          mockContext = runInMockContext;
        }
        if (typeof runInMockContext === 'function') {
          mockContext = runInMockContext();
        }
        // @ts-ignore
        merge(global.window, mockContext || {});
      }
    };
  }
  return () => {
    // noop
  };
};
