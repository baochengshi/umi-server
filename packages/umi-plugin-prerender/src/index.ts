import { IApi } from 'umi-types';
import * as path from 'path';
import * as fs from 'fs';
import * as mkdirp from 'mkdirp';
// @ts-ignore
import { IConfig } from 'umi-server/lib/index';
import { getStaticRoutePaths, fixHtmlSuffix, findJSON, isDynamicRoute, routeToFile } from './utils';

export interface IOpts extends IConfig {
  include?: string[];
  exclude?: string[];
  /** disable ssr BOM polyfill */
  disablePolyfill?: boolean;
  // htmlSuffix
  htmlSuffix?: boolean;
  runInMockContext?: object | (() => object);
}

export default (api: IApi, opts: IOpts) => {
  const { debug, config, findJS, log } = api;
  const {
    include = [],
    exclude = [],
    htmlSuffix = false,
    disablePolyfill = false,
    runInMockContext = {},
    ...restOpts
  } = opts || {};
  if (!config.ssr) {
    throw new Error('config must use { ssr: true } when using umi preRender plugin');
  }

  api.onPatchRoute(({ route }) => {
    debug(`route before, ${JSON.stringify(route)}`);
    if (htmlSuffix) {
      fixHtmlSuffix(route);
    }
    debug(`route after, ${JSON.stringify(route)}`);
  });

  // onBuildSuccess hook
  api.onBuildSuccessAsync(async () => {
    const { routes, _ } = api;
    const { absOutputPath } = api.paths;
    const { manifestFileName = 'ssr-client-mainifest.json' } = config.ssr as any;

    // require serverRender function
    const defaultHtmlTemplate = fs.readFileSync(path.join(absOutputPath, 'index.html'), 'utf-8');
    const filename = findJS(absOutputPath, 'umi.server');
    const manifest = findJSON(absOutputPath, manifestFileName);
    if (!filename) {
      throw new Error("can't find umi.server.js file");
    }
    const server = require('umi-server');
    const render = server({
      filename,
      manifest,
      polyfill: !disablePolyfill,
      ...restOpts,
    });

    const routePaths: string[] = getStaticRoutePaths(_, routes);

    // get render paths
    const renderPaths = routePaths
      .filter(path => (include?.length > 0 ? include.includes(path) : true))
      .filter(path => !exclude.includes(path));
    debug(`renderPaths: ${renderPaths.join(',')}`);
    log.start('umiJS prerender start');
    // loop routes
    for (const url of renderPaths) {
      const ctx = {
        url,
        req: {
          url,
        },
        request: {
          url,
        },
      };

      let ssrHtml = defaultHtmlTemplate;

      // 动态路由走默认 default.html
      if (!isDynamicRoute(url)) {
        const serverRenderRes = await render(ctx, {
          runInMockContext,
        });
        ssrHtml = serverRenderRes?.ssrHtml || defaultHtmlTemplate;
      }

      const filename = routeToFile(url);
      try {
        // write html file
        const outputRoutePath = path.join(absOutputPath, filename);
        const dir = path.join(absOutputPath, filename.substring(0, filename.lastIndexOf('/')));
        mkdirp.sync(dir);
        fs.writeFileSync(outputRoutePath, ssrHtml);
        log.complete(`${path.join(filename, filename)}`);
      } catch (e) {
        log.fatal(`${url} render ${filename} failed`, e);
      }
    }
    log.success('umiJS prerender success!');
  });
};

/** export routeToFile logic */
export { routeToFile };
