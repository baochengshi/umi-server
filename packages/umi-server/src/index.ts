/* eslint-disable import/no-dynamic-require */
import { join } from 'path';
import { load } from 'cheerio';
import React from 'react';
import compose from './compose';
import _log from './debug';
import {
  nodePolyfillDecorator,
  injectChunkMaps,
  _getDocumentHandler,
  filterRootContainer,
} from './utils';

interface IDynamicChunkMap {
  js: string[];
  css: string[];
}

export interface IFilterContext {
  publicPath: string;
}

export interface IArgs extends Partial<IFilterContext> {
  chunkMap: IDynamicChunkMap;
}
type cheerio = ReturnType<typeof load>;
export type IHandler = ($: cheerio, args: IArgs) => cheerio;

export interface IPolyfill {
  host?: string;
}

type IContextFunc = () => object;

export interface IConfig {
  /** prefix path for `filename` and `manifest`, if both in the same directory */
  root?: string;
  /** ssr manifest, default: `${root}/ssr-client-mainifest.json` */
  manifest?: string;
  /** umi ssr server file, default: `${root}/umi.server.js` */
  filename?: string;
  /** default false */
  polyfill?: boolean | IPolyfill;
  /** use renderToStaticMarkup  */
  staticMarkup?: boolean;
  /** replace the default ReactDOMServer.renderToString */
  customRender?: (args: IRenderArgs) => Promise<string>;
  /** handler function for user to modify render html accounding cheerio */
  postProcessHtml?: IHandler | IHandler[];
  /** is dev env, default NODE_ENV=development */
  dev?: boolean;
  /** TODO: serverless */
  serverless?: boolean;
}

export interface IRenderOpts extends Pick<IConfig, 'polyfill'> {
  /** mock global object like { g_lang: 'zh-CN' } => global.window.g_lang / global.g_lang  */
  runInMockContext?: object | IContextFunc;
}

export interface IRenderArgs {
  htmlElement: React.ReactNode;
  rootContainer: React.ReactNode;
  matchPath: string;
  g_initialData: any;
  chunkMap?: IDynamicChunkMap;
}

export interface IContext {
  req: {
    url: string;
  };
}

export interface IResult {
  ssrHtml: string;
  matchPath: string;
  chunkMap?: IDynamicChunkMap;
}

export type IServer = (
  config: IConfig,
) => (ctx: IContext, renderOpts?: IRenderOpts) => Promise<IResult>;

const server: IServer = config => {
  const {
    root,
    manifest = join(root, 'ssr-client-mainifest.json'),
    filename = join(root, 'umi.server'),
    staticMarkup = false,
    polyfill = false,
    postProcessHtml = $ => $,
    customRender,
    dev = process.env.NODE_ENV === 'development',
  } = config;
  const polyfillHost =
    typeof polyfill === 'object' && polyfill.host ? polyfill.host : 'http://localhost';
  const nodePolyfill = nodePolyfillDecorator(!!polyfill, polyfillHost);
  const serverRender = require(filename);
  const manifestFile = require(manifest);
  if (dev) {
    // remove module cache when in dev mode
    delete require.cache[require.resolve(filename)];
    delete require.cache[require.resolve(manifest)];
  }
  const { ReactDOMServer } = serverRender;

  _log('manifestFile', _log);

  return async (ctx, renderOpts = {}) => {
    const {
      req: { url },
    } = ctx;
    // polyfill pathname
    nodePolyfill(renderOpts, {
      url,
    });
    const serverRenderRes: Omit<IRenderArgs, 'chunkMap'> = await serverRender.default(ctx);
    const { htmlElement, matchPath, g_initialData } = serverRenderRes;
    // if not found, return undefined
    if (!matchPath) {
      return {
        ssrHtml: undefined,
        matchPath: undefined,
        chunkMap: { js: [], css: [] },
        g_initialData: {},
      };
    }
    const chunkMap: IDynamicChunkMap = manifestFile[matchPath];
    const reactRender = ReactDOMServer[staticMarkup ? 'renderToStaticMarkup' : 'renderToString'];

    const renderString: string =
      typeof customRender === 'function'
        ? await customRender({
            ...serverRenderRes,
            chunkMap,
          })
        : reactRender(htmlElement);

    const processHtmlHandlers = Array.isArray(postProcessHtml)
      ? postProcessHtml
      : [postProcessHtml];
    const composeRender = compose(
      injectChunkMaps,
      // user define handler
      ...processHtmlHandlers,
    );
    const ssrHtml = filterRootContainer(renderString, (layoutHtml, context) => {
      const $ = _getDocumentHandler(layoutHtml);
      const handlerOpts = {
        ...context,
        chunkMap,
      };
      // compose all layoutHtml handlers
      return composeRender($, handlerOpts).html();
    });

    // enable render rootContainer
    // const ssrHtmlElement =
    return {
      ssrHtml,
      matchPath,
      chunkMap,
      g_initialData,
    };
  };
};

export default server;
