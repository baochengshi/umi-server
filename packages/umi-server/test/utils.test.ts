import { filterRootContainer } from '../src/utils';

describe('utils', () => {
  it('filterRootContainer normal head', () => {
    const beforeHtml = `
      <!DOCTYPE html><html data-reactroot><head><link rel="stylesheet" href="/"><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no"><script>window.g_useSSR=true;
    window.g_initialData = {"data":{"ssr":"http:\u002F\u002F127.0.0.1:7001","csr":"http:\u002F\u002F127.0.0.1:8000"}};</script><script>window.routerBase = "/";</script></head><body><div id="root"><div class="normal___1KW4T"><h1>Page index</h1><h2>csr: <!-- -->http://127.0.0.1:8000</h2></div></div><script src="/umi.js"></script></body></html>`;
    const newHtml = filterRootContainer(beforeHtml, html => {
      return html.replace('</head>', '<script>console.log(1);</script></head>');
    });
    expect(newHtml).toMatchSnapshot();
  });

  it('filterRootContainer normal body', () => {
    const beforeHtml = `
      <!DOCTYPE html><html data-reactroot><head><link rel="stylesheet" href="/"><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no"><script>window.g_useSSR=true;
    window.g_initialData = {"data":{"ssr":"http:\u002F\u002F127.0.0.1:7001","csr":"http:\u002F\u002F127.0.0.1:8000"}};</script><script>window.routerBase = "/";</script></head><body><div id="root"><div class="normal___1KW4T"><h1>Page index</h1><h2>csr: <!-- -->http://127.0.0.1:8000</h2></div></div><script src="/umi.js"></script></body></html>`;
    const newHtml = filterRootContainer(beforeHtml, html => {
      return html.replace('</body>', '<script>console.log(1);</script></body>');
    });
    expect(newHtml).toMatchSnapshot();
  });
});
