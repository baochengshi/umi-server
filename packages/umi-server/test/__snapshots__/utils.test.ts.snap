// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils filterRootContainer normal body 1`] = `
"
      <!DOCTYPE html><html data-reactroot><head><link rel=\\"stylesheet\\" href=\\"/\\"><meta charset=\\"utf-8\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no\\"><script>window.g_useSSR=true;
    window.g_initialData = {\\"data\\":{\\"ssr\\":\\"http://127.0.0.1:7001\\",\\"csr\\":\\"http://127.0.0.1:8000\\"}};</script><script>window.routerBase = \\"/\\";</script></head><body><div id=\\"root\\"><div class=\\"normal___1KW4T\\"><h1>Page index</h1><h2>csr: <!-- -->http://127.0.0.1:8000</h2></div></div><script src=\\"/umi.js\\"></script><script>console.log(1);</script></body></html>"
`;

exports[`utils filterRootContainer normal head 1`] = `
"
      <!DOCTYPE html><html data-reactroot><head><link rel=\\"stylesheet\\" href=\\"/\\"><meta charset=\\"utf-8\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no\\"><script>window.g_useSSR=true;
    window.g_initialData = {\\"data\\":{\\"ssr\\":\\"http://127.0.0.1:7001\\",\\"csr\\":\\"http://127.0.0.1:8000\\"}};</script><script>window.routerBase = \\"/\\";</script><script>console.log(1);</script></head><body><div id=\\"root\\"><div class=\\"normal___1KW4T\\"><h1>Page index</h1><h2>csr: <!-- -->http://127.0.0.1:8000</h2></div></div><script src=\\"/umi.js\\"></script></body></html>"
`;
